src/components/structures/RoomView.js

```diff
@@ -618,10 +619,27 @@ module.exports = React.createClass({
         }
     },
 
-    _updatePreviewUrlVisibility: function(room) {
-        this.setState({
-            showUrlPreview: SettingsStore.getValue("urlPreviewsEnabled", room.roomId),
-        });
+    _updatePreviewUrlVisibility: function({roomId}) {
+        const levels = [
+            SettingLevel.ROOM_DEVICE,
+            SettingLevel.ROOM_ACCOUNT,
+        ];
+        let showUrlPreview;
+        // in e2ee rooms only care about room-device and room-account, so that user has to explicitly enable previews
+        if (MatrixClientPeg.get().isRoomEncrypted(roomId)) {
+            for (const level of levels) {
+                const value = SettingsStore.getValueAt(level, "urlPreviewsEnabled", roomId, true, true);
+                if (value === Boolean(value)) { // if is Boolean
+                    showUrlPreview = value;
+                    break;
+                }
+            }
```

Comment:
Hmm, this feels like a bit of an abuse of the Settings API in terms of relying a setting existing at particular levels to enable something rather than using it consistently.

Moreover, is there not a privacy leak here still whereby if I enable URL previews in a room (for me) and then the room becomes an e2e room, I'll still be requesting URL previews? Maybe what we want here is an explicit, per-room, "enable URL previews despite e2e" flag.

Reply:
That is a very great point, so the previews must be enabled post-e2e-being-enabled